lib/Monads/wp/WPC.thy

```diff
@@ -213,67 +213,88 @@ end;
 val _ =
     Outer_Syntax.command
         @{command_keyword "wpc_setup"}
-        "Add wpc stuff"
+        "Add new WPC term and helper rule"
         (P.term -- P.name >> (fn (tm, thm) => Toplevel.local_theory NONE NONE (add_wpc tm thm)))
 
 end;
 end;
-
 \<close>
 
 ML \<open>
-
-val wp_cases_tactic_weak = WeakestPreCases.wp_cases_tac @{thms wpc_weak_processors};
+val wp_cases_tactic_weak   = WeakestPreCases.wp_cases_tac @{thms wpc_weak_processors};
 val wp_cases_method_strong = WeakestPreCases.wp_cases_method @{thms wpc_processors};
 val wp_cases_method_weak   = WeakestPreCases.wp_cases_method @{thms wpc_weak_processors};
 val wp_cases_method_vweak  = WeakestPreCases.wp_cases_method @{thms wpc_vweak_processors};
-
 \<close>
 
+(* Main proof methods: *)
 method_setup wpc0 = \<open>wp_cases_method_strong\<close>
   "case splitter for weakest-precondition proofs"
 
 method_setup wpcw0 = \<open>wp_cases_method_weak\<close>
   "weak-form case splitter for weakest-precondition proofs"
 
+(* Instances specifically for wp (introducing schematic guards automatically): *)
 method wpc = (wp_pre, wpc0)
 method wpcw = (wp_pre, wpcw0)
 
-definition
-  wpc_test :: "'a set \<Rightarrow> ('a \<times> 'b) set \<Rightarrow> 'b set \<Rightarrow> bool"
-  where
- "wpc_test P R S \<equiv> (R `` P) \<subseteq> S"
+(* Test and example *)
+experiment
+begin
 
+(* Assume some kind of judgement wpc_test with a precondition P of type set and a
+   precondition Q of type 'a \<Rightarrow> bool: *)
+definition wpc_test :: "'a set \<Rightarrow> ('a \<Rightarrow> bool) \<Rightarrow> ('a \<times> 'b) set \<Rightarrow> 'b set \<Rightarrow> bool" where
+  "wpc_test P Q R S \<equiv> (R `` P) \<subseteq> S"
+
+(* Weakening rule to introduce schematics for the two guards *)
 lemma wpc_test_weaken:
-  "\<lbrakk> wpc_test Q R S; P \<subseteq> Q \<rbrakk> \<Longrightarrow> wpc_test P R S"
+  "\<lbrakk> wpc_test Q X' R S; P \<subseteq> Q; \<And>s. X s \<Longrightarrow> X' s \<rbrakk> \<Longrightarrow> wpc_test P X R S"
   by (simp add: wpc_test_def, blast)
 
-lemma wpc_helper_validF:
-  "wpc_test Q' R S \<Longrightarrow> wpc_helper (P, P') (Q, Q') (wpc_test P' R S)"
-  by (simp add: wpc_test_def wpc_helper_def, blast)
+(* Setup rule, establishes connection between wpc_helper and judgment wpc_test.
+   The precondition has the judgement with transformed (Q) guards, the conclusion has the
+   helper predicate with the judgement applied to the original (P) guards. The guard triple
+   arguments of wpc_helper must be in the form below (no arguments or patterns) for the method
```

Comment:
"The guard triple arguments of ..." - not comprehending what this means

Reply:
To be honest, I don't really comprehend this any more either. Might have lost a word somewhere in editing that sentence. Will adjust.